### PR TITLE
Remove leading ../ from the asset .flx zip map and from the font manifest json

### LINKED ISSUE #8403: Asset paths with .. in them cause directories to be created in weird places.
This was specifically happening with font assets.  A font folder was unexpectedly being created as a sibling to the build directly.

### PR DESCRIPTION
**Pull request by xster**
Preserves the DevFSFileContent pointing to asset files in the .yaml file's parents

Fixes #8403

```
[{"fonts":[{"asset":"../../apps/fonts/third_party/roboto/Roboto-Regular.ttf"},{"style":"italic","asset":"../../apps/fonts/third_party/roboto/Roboto-Italic.ttf"},{"weight":100,"asset":"../../apps/fonts/third_party/roboto/Roboto-Thin.ttf"},{"style":"italic","weight":100,"asset":"../../apps/fonts/third_party/roboto/Roboto-ThinItalic.ttf"},{"weight":300,"asset":"../../apps/fonts/third_party/roboto/Roboto-Light.ttf"},{"style":"italic","weight":300,"asset":"../../apps/fonts/third_party/roboto/Roboto-LightItalic.ttf"},{"weight":500,"asset":"../../apps/fonts/third_party/roboto/Roboto-Medium.ttf"},{"style":"italic","weight":500,"asset":"../../apps/fonts/third_party/roboto/Roboto-MediumItalic.ttf"},{"weight":700,"asset":"../../apps/fonts/third_party/roboto/Roboto-Bold.ttf"},{"style":"italic","weight":700,"asset":"../../apps/fonts/third_party/roboto/Roboto-BoldItalic.ttf"},{"weight":900,"asset":"../../apps/fonts/third_party/roboto/Roboto-Black.ttf"},{"style":"italic","weight":900,"asset":"../../apps/fonts/third_party/roboto/Roboto-BlackItalic.ttf"}],"family":"Roboto"},{"fonts":[{"asset":"../../apps/fonts/third_party/robotocondensed/RobotoCondensed-Regular.ttf"},{"style":"italic","asset":"../../apps/fonts/third_party/robotocondensed/RobotoCondensed-Italic.ttf"},{"weight":300,"asset":"../../apps/fonts/third_party/robotocondensed/RobotoCondensed-Light.ttf"},{"style":"italic","weight":300,"asset":"../../apps/fonts/third_party/robotocondensed/RobotoCondensed-LightItalic.ttf"},{"weight":700,"asset":"../../apps/fonts/third_party/robotocondensed/RobotoCondensed-Bold.ttf"},{"style":"italic","weight":700,"asset":"../../apps/fonts/third_party/robotocondensed/RobotoCondensed-BoldItalic.ttf"}],"family":"RobotoCondensed"}]
```

becomes 

```
[{"fonts":[{"asset":"fonts/third_party/roboto/Roboto-Regular.ttf"},{"style":"italic","asset":"fonts/third_party/roboto/Roboto-Italic.ttf"},{"weight":100,"asset":"fonts/third_party/roboto/Roboto-Thin.ttf"},{"style":"italic","weight":100,"asset":"fonts/third_party/roboto/Roboto-ThinItalic.ttf"},{"weight":300,"asset":"fonts/third_party/roboto/Roboto-Light.ttf"},{"style":"italic","weight":300,"asset":"fonts/third_party/roboto/Roboto-LightItalic.ttf"},{"weight":500,"asset":"fonts/third_party/roboto/Roboto-Medium.ttf"},{"style":"italic","weight":500,"asset":"fonts/third_party/roboto/Roboto-MediumItalic.ttf"},{"weight":700,"asset":"fonts/third_party/roboto/Roboto-Bold.ttf"},{"style":"italic","weight":700,"asset":"fonts/third_party/roboto/Roboto-BoldItalic.ttf"},{"weight":900,"asset":"fonts/third_party/roboto/Roboto-Black.ttf"},{"style":"italic","weight":900,"asset":"fonts/third_party/roboto/Roboto-BlackItalic.ttf"}],"family":"Roboto"},{"fonts":[{"asset":"fonts/third_party/robotocondensed/RobotoCondensed-Regular.ttf"},{"style":"italic","asset":"fonts/third_party/robotocondensed/RobotoCondensed-Italic.ttf"},{"weight":300,"asset":"fonts/third_party/robotocondensed/RobotoCondensed-Light.ttf"},{"style":"italic","weight":300,"asset":"fonts/third_party/robotocondensed/RobotoCondensed-LightItalic.ttf"},{"weight":700,"asset":"fonts/third_party/robotocondensed/RobotoCondensed-Bold.ttf"},{"style":"italic","weight":700,"asset":"fonts/third_party/robotocondensed/RobotoCondensed-BoldItalic.ttf"}],"family":"RobotoCondensed"}]
```

### CODE REVIEW

**Comment by xster** [comment]
Will add tests if this is sane

File: packages/flutter_tools/lib/src/asset.dart [hunk 1]
```diff
@@ -317,7 +319,9 @@ DevFSContent _createFontManifest(Map<String, dynamic> manifestDescriptor,
     fonts.addAll(manifestDescriptor['fonts']);
   if (fonts.isEmpty)
     return null;
-  return new DevFSStringContent(JSON.encode(fonts));
+  String json = JSON.encode(fonts);
+  json = json.replaceAll(_assetJsonLeadingParentRegExp, '"asset":"');
```
Comment (abarth) [inline packages/flutter_tools/lib/src/asset.dart]: I don't think we should be running regular expressions over the encoded JSON.  Can we resolve the issue while we still have the object model?

File: packages/flutter_tools/lib/src/asset.dart [hunk 2]
```diff
@@ -401,6 +405,8 @@ Map<_Asset, List<_Asset>> _parseAssets(
   return result;
 }
 
+final RegExp _leadingParentRegExp = new RegExp(r'(\.\.\/)+');
```
Comment (abarth) [inline packages/flutter_tools/lib/src/asset.dart]: This doesn't handle the case of "foo/../../../../bar/baz"

Does https://www.dartdocs.org/documentation/path/1.4.1/path/normalize.html not do what we want?  Maybe we need to resolve it against a base path of `/` or something in addition.

File: packages/flutter_tools/lib/src/asset.dart (hunk 1 above)
Comment (xster) [inline packages/flutter_tools/lib/src/asset.dart]: Ya, I struggled on this too and puked some blood as I clicked git push...
Though the problem is manifestDescriptor (a bit above) is a YamlMap that's unmodifiable and might have some particular bevahiours. Either way some conversion needed to be done before modifying it.

Comment (abarth) [inline packages/flutter_tools/lib/src/asset.dart]: Maybe we should convert it to some mutable representation, mutate it, and then serialize it out to JSON

Comment (xster) [inline packages/flutter_tools/lib/src/asset.dart]: ya, I started down that path initially with some graph traversal though I would need to handle some conversion edge cases in the behaviour that normal dart maps don't support like using NaN as map keys etc that seems to get out of hand quickly. 

I don't suppose there are easy ways in Dart to unroll mixins that are bringing in the unmodifiable behaviours?

Comment (abarth) [inline packages/flutter_tools/lib/src/asset.dart]: I wouldn't try to make a general converter.  I'd just handle the specific schema we expect for fonts.

Comment (xster) [inline packages/flutter_tools/lib/src/asset.dart]: Created another pr for option 2
#8943

Not sure how I feel about creating a half generic converter though. Feels like there are more assumptions in the converter than in the json string manipulator.

Created https://github.com/dart-lang/yaml/issues/26 which I would feel more comfortable in the long run.

**Comment by Hixie** [comment]
@xster is this PR still live?

**Comment by xster** [comment]
I'm keeping my local branch and closing this for now

**Pull request closed without merging**